wicker/core/datasets.py

```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
         pass
 
     @abc.abstractmethod
-    def schema(self) -> DatasetSchema:
-        """Return the schema of the dataset."""
+    def arrow_table(self) -> pyarrow.Table:
+        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
         pass
 
     @abc.abstractmethod
-    def arrow_table(self) -> pyarrow.Table:
-        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
+    def schema(self) -> DatasetSchema:
+        """Return the schema of the dataset."""
         pass
 
 
-class S3Dataset(AbstractDataset):
-    """Implementation for a Map-based dataset"""
+class BaseDataset(AbstractDataset):
+    """Provides an implementation for part of the AbstractDataset interface for datasets whose items
+    and schema can be read from AbstractDataStorage."""
 
     def __init__(
         self,
         dataset_name: str,
         dataset_version: str,
         dataset_partition_name: str,
-        local_cache_path_prefix: str = os.getenv("TMPDIR", "/tmp"),
+        pa_filesystem: pafs.FileSystem,
```

Comment:
Curious with the relationship of this to the DataStorage class, seems like concept overlap

Reply:
Yes, of course the storage type should determine the arrow file system type. Maybe sometime over good coffee I can explain the history with my partner on the project. He was smart and fast programmer (now he is cofounder of the company "Daft Dataframe"), but at least then he was super wild. I did not get to review this code.

But back to the issue. Since the current `S3Dataset` [constructor](https://github.com/woven-planet/wicker/blob/main/wicker/core/datasets.py#L150) accepts a `pa_filesystem` object, I'm not sure if we can fix the issue or not since we need to keep backwards compatibility with the current code.

Here in `BaseDataset` constructor, we could remove `pa_filesystem` from the `__init__` parameters and set it as a member variable by calling a new function `self._pa_filesystem = self._storage.arrow_file_system()` that we add to the `AbstractStorage` class. Then to maintain backwards compatibility in the `S3Dataset` constructor, we can just leave the current logic [here](https://github.com/woven-planet/wicker/blob/main/wicker/core/datasets.py#L186) and override the `self._pa_filesystem` member variable. That is my best idea - let know what you think!

Reply:
Thanks for the explanation, I know this is not introduced by your, but from the concept, the original wicker author reinvent the datastorage class but hard coded the abstract with arrow lib. Really did not get their point of the design, make me really confusion.
No matter what they are doing, at least at that time, not good programmer and did not fix the tech debt

Reply:
We can do this refactor after this PR